crates/router/src/core/payment_methods.rs

```diff
@@ -1656,6 +1670,159 @@ async fn generate_saved_pm_response(
+        )
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::PaymentMethodNotFound)?;
+
+    let pmd: Option<api::PaymentMethodResponseData> = Some(
+        cards::retrieve_payment_method_from_vault(
+            &state,
+            &merchant_account,
+            &payment_method.customer_id,
+            &payment_method,
+        )
+        .await
+        .attach_printable("Failed to retrieve payment method from vault")?
+        .data
+        .parse_struct("PaymentMethodCreateData")
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to parse PaymentMethodCreateData")?,
+    )
+    .map(|pmd| match pmd {
+        api::PaymentMethodCreateData::Card(card) => {
+            api::PaymentMethodResponseData::Card(card.into())
+        }
+    });
+
+    let resp = api::PaymentMethodResponse {
+        merchant_id: payment_method.merchant_id.to_owned(),
+        customer_id: payment_method.customer_id.to_owned(),
+        payment_method_id: payment_method.id.get_string_repr(),
+        payment_method: payment_method.payment_method,
+        payment_method_type: payment_method.payment_method_type,
+        metadata: payment_method.metadata.clone(),
+        created: Some(payment_method.created_at),
+        recurring_enabled: false,
+        last_used_at: Some(payment_method.last_used_at),
+        client_secret: payment_method.client_secret.clone(),
+        payment_method_data: pmd,
+    };
+
+    Ok(services::ApplicationResponse::Json(resp))
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub async fn update_payment_method(
+    state: SessionState,
+    merchant_account: domain::MerchantAccount,
+    req: api::PaymentMethodUpdate,
+    payment_method_id: &str,
+    key_store: domain::MerchantKeyStore,
+) -> RouterResponse<api::PaymentMethodResponse> {
+    let db = state.store.as_ref();
+
+    let pm_id = id_type::GlobalPaymentMethodId::generate_from_string(payment_method_id.to_string())
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Unable to generate GlobalPaymentMethodId")?;
+
+    let payment_method = db
+        .find_payment_method(
+            &((&state).into()),
+            &key_store,
+            &pm_id,
+            merchant_account.storage_scheme,
+        )
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::PaymentMethodNotFound)?;
+    let current_vault_id = payment_method.locker_id.clone();
+
+    when(
+        payment_method.status == enums::PaymentMethodStatus::AwaitingData,
+        || {
+            Err(errors::ApiErrorResponse::InvalidRequestData {
+                message: "This Payment method is awaiting data and hence cannot be updated"
+                    .to_string(),
+            })
+        },
+    )?;
+
+    let pmd: api::PaymentMethodCreateData = cards::retrieve_payment_method_from_vault(
```

Comment:
Why we are constructing CreateData back for update? And rename `PaymentMethodCreateData` to `PaymentMethodData`

Reply:
Have created a new Vaulting Type, should be used specifically for Vaulting.